File: src/main/java/org/moinex/service/BondService.java
/*
 * Filename: BondService.java
 * Created on: January  2, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.service;

import jakarta.persistence.EntityExistsException;
import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.moinex.model.Category;
import org.moinex.model.investment.Bond;
import org.moinex.model.investment.BondOperation;
import org.moinex.model.wallettransaction.WalletTransaction;
import org.moinex.repository.investment.BondOperationRepository;
import org.moinex.repository.investment.BondRepository;
import org.moinex.util.enums.BondType;
import org.moinex.util.enums.InterestIndex;
import org.moinex.util.enums.InterestType;
import org.moinex.util.enums.OperationType;
import org.moinex.util.enums.TransactionStatus;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
public class BondService {

    private final BondRepository bondRepository;
    private final BondOperationRepository bondOperationRepository;
    private final WalletTransactionService walletTransactionService;
    private final WalletService walletService;

    @Autowired
    public BondService(
            BondRepository bondRepository,
            BondOperationRepository bondOperationRepository,
            WalletTransactionService walletTransactionService,
            WalletService walletService) {
        this.bondRepository = bondRepository;
        this.bondOperationRepository = bondOperationRepository;
        this.walletTransactionService = walletTransactionService;
        this.walletService = walletService;
    }

    @Transactional
    public void addBond(
            String name,
            String symbol,
            BondType bondType,
            String issuer,
            LocalDateTime maturityDate,
            InterestType interestType,
            InterestIndex interestIndex,
            BigDecimal interestRate) {

        if (symbol != null && !symbol.isBlank() && bondRepository.existsBySymbol(symbol)) {
            throw new EntityExistsException("Bond with symbol " + symbol + " already exists");
        }

        Bond bond =
                Bond.builder()
                        .name(name)
                        .symbol(symbol)
                        .type(bondType)
                        .issuer(issuer)
                        .maturityDate(maturityDate)
                        .interestType(interestType)
                        .interestIndex(interestIndex)
                        .interestRate(interestRate)
                        .archived(false)
                        .build();

        bondRepository.save(bond);
        log.info("Bond {} added successfully", name);
    }

    @Transactional(readOnly = true)
    public List<Bond> getAllNonArchivedBonds() {
        return bondRepository.findByArchivedFalseOrderByNameAsc();
    }

    @Transactional(readOnly = true)
    public List<Bond> getAllArchivedBonds() {
        return bondRepository.findByArchivedTrueOrderByNameAsc();
    }

    @Transactional(readOnly = true)
    public Bond getBondById(Integer id) {
        return bondRepository
                .findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Bond not found with id: " + id));
    }

    @Transactional
    public void updateBond(
            Integer id,
            String name,
            String symbol,
            BondType bondType,
            String issuer,
            LocalDateTime maturityDate,
            InterestType interestType,
            InterestIndex interestIndex,
            BigDecimal interestRate) {

        Bond bond =
                bondRepository
                        .findById(id)
                        .orElseThrow(
                                () -> new EntityNotFoundException("Bond not found with id: " + id));

        bond.setName(name);
        bond.setSymbol(symbol);
        bond.setType(bondType);
        bond.setIssuer(issuer);
        bond.setMaturityDate(maturityDate);
        bond.setInterestType(interestType);
        bond.setInterestIndex(interestIndex);
        bond.setInterestRate(interestRate);

        bondRepository.save(bond);
        log.info("Bond {} updated successfully", name);
    }

    @Transactional(readOnly = true)
    public Integer getOperationCountByBond(Integer bondId) {
        Bond bond =
                bondRepository
                        .findById(bondId)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                "Bond not found with id: " + bondId));

        return bondOperationRepository.findByBondOrderByOperationDateAsc(bond).size();
    }

    @Transactional
    public void deleteBond(Integer id) {
        Bond bond =
                bondRepository
                        .findById(id)
                        .orElseThrow(
                                () -> new EntityNotFoundException("Bond not found with id: " + id));

        // Check if the bond has operations associated with it
        List<BondOperation> operations =
                bondOperationRepository.findByBondOrderByOperationDateAsc(bond);
        if (!operations.isEmpty()) {
            throw new IllegalStateException(
                    String.format(
                            "Bond with id %d has operations associated with it and cannot be"
                                    + " deleted. Remove the operations first or archive the bond",
                            id));
        }

        bondRepository.delete(bond);
        log.info("Bond {} was permanently deleted", bond.getName());
    }

    @Transactional
    public void archiveBond(Integer id) {
        Bond bond =
                bondRepository
                        .findById(id)
                        .orElseThrow(
                                () -> new EntityNotFoundException("Bond not found with id: " + id));

        bond.setArchived(true);
        bondRepository.save(bond);
        log.info("Bond {} archived", bond.getName());
    }

    @Transactional
    public void unarchiveBond(Integer id) {
        Bond bond =
                bondRepository
                        .findById(id)
                        .orElseThrow(
                                () -> new EntityNotFoundException("Bond not found with id: " + id));

        bond.setArchived(false);
        bondRepository.save(bond);
        log.info("Bond {} unarchived", bond.getName());
    }

    @Transactional(readOnly = true)
    public BigDecimal getCurrentQuantity(Bond bond) {
        List<BondOperation> operations =
                bondOperationRepository.findByBondOrderByOperationDateAsc(bond);

        BigDecimal quantity = BigDecimal.ZERO;
        for (BondOperation op : operations) {
            if (op.getOperationType() == OperationType.BUY) {
                quantity = quantity.add(op.getQuantity());
            } else {
                quantity = quantity.subtract(op.getQuantity());
            }
        }
        return quantity;
    }

    @Transactional(readOnly = true)
    public BigDecimal getAverageUnitPrice(Bond bond) {
        List<BondOperation> purchases =
                bondOperationRepository.findByBondAndOperationTypeOrderByOperationDateAsc(
                        bond, OperationType.BUY);

        if (purchases.isEmpty()) {
            return BigDecimal.ZERO;
        }

        BigDecimal totalValue = BigDecimal.ZERO;
        BigDecimal totalQuantity = BigDecimal.ZERO;

        for (BondOperation purchase : purchases) {
            totalValue = totalValue.add(purchase.getUnitPrice().multiply(purchase.getQuantity()));
            totalQuantity = totalQuantity.add(purchase.getQuantity());
        }

        if (totalQuantity.compareTo(BigDecimal.ZERO) == 0) {
            return BigDecimal.ZERO;
        }

        return totalValue.divide(totalQuantity, 8, java.math.RoundingMode.HALF_UP);
    }

    @Transactional
    public void addOperation(
            Integer bondId,
            Integer walletId,
            OperationType operationType,
            BigDecimal quantity,
            BigDecimal unitPrice,
            LocalDate operationDate,
            BigDecimal fees,
            BigDecimal taxes,
            BigDecimal netProfit,
            Category category,
            String description,
            TransactionStatus status) {
        Bond bond =
                bondRepository
                        .findById(bondId)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                "Bond not found with id: " + bondId));

        if (quantity.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Quantity must be greater than zero");
        }

        if (unitPrice.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Unit price must be greater than zero");
        }

        if (operationType == OperationType.SELL) {
            BigDecimal currentQuantity = getCurrentQuantity(bond);
            if (currentQuantity.compareTo(quantity) < 0) {
                throw new IllegalArgumentException(
                        "Insufficient bond quantity. Available: "
                                + currentQuantity
                                + ", Requested: "
                                + quantity);
            }
        }

        BigDecimal feesAmount = fees != null ? fees : BigDecimal.ZERO;
        BigDecimal taxesAmount = taxes != null ? taxes : BigDecimal.ZERO;
        BigDecimal netProfitAmount = netProfit != null ? netProfit : BigDecimal.ZERO;

        BigDecimal baseAmount = unitPrice.multiply(quantity);
        BigDecimal amount;

        if (operationType == OperationType.BUY) {
            amount = baseAmount.add(feesAmount).add(taxesAmount);
        } else {
            amount = baseAmount.add(netProfitAmount);
        }

        LocalDateTime dateTime = operationDate.atStartOfDay();
        Integer transactionId;

        if (operationType == OperationType.BUY) {
            transactionId =
                    walletTransactionService.addExpense(
                            walletId, category, dateTime, amount, description, status);
        } else {
            transactionId =
                    walletTransactionService.addIncome(
                            walletId, category, dateTime, amount, description, status);
        }

        WalletTransaction walletTransaction =
                walletTransactionService.getTransactionById(transactionId);

        BondOperation operation =
                BondOperation.builder()
                        .bond(bond)
                        .operationType(operationType)
                        .quantity(quantity)
                        .unitPrice(unitPrice)
                        .fees(fees)
                        .taxes(taxes)
                        .netProfit(netProfit)
                        .walletTransaction(walletTransaction)
                        .build();

        bondOperationRepository.save(operation);

        log.info(
                "BondOperation {} with id {} added to bond {}. Wallet transaction with id {}"
                        + " created",
                operationType,
                operation.getId(),
                bond.getName(),
                transactionId);
    }

    @Transactional(readOnly = true)
    public List<BondOperation> getAllOperations() {
        return bondOperationRepository.findAllByOrderByOperationDateDesc();
    }

    @Transactional(readOnly = true)
    public List<BondOperation> getOperationsByBond(Bond bond) {
        return bondOperationRepository.findByBondOrderByOperationDateAsc(bond);
    }

    @Transactional(readOnly = true)
    public BigDecimal calculateProfit(Bond bond) {
        List<BondOperation> operations =
                bondOperationRepository.findByBondOrderByOperationDateAsc(bond);

        BigDecimal totalProfit = BigDecimal.ZERO;

        for (BondOperation op : operations) {
            if (op.getOperationType() == OperationType.SELL && op.getNetProfit() != null) {
                totalProfit = totalProfit.add(op.getNetProfit());
            }
        }

        return totalProfit;
    }

    @Transactional(readOnly = true)
    public BigDecimal getTotalInvestedValue() {
        List<Bond> bonds = bondRepository.findByArchivedFalseOrderByNameAsc();
        BigDecimal total = BigDecimal.ZERO;

        for (Bond bond : bonds) {
            BigDecimal currentQuantity = getCurrentQuantity(bond);
            if (currentQuantity.compareTo(BigDecimal.ZERO) > 0) {
                BigDecimal averagePrice = getAverageUnitPrice(bond);
                total = total.add(averagePrice.multiply(currentQuantity));
            }
        }

        return total;
    }

    @Transactional(readOnly = true)
    public BigDecimal getTotalCurrentValue(BigDecimal currentMarketPrice) {
        List<Bond> bonds = bondRepository.findByArchivedFalseOrderByNameAsc();
        BigDecimal total = BigDecimal.ZERO;

        for (Bond bond : bonds) {
            total = total.add(currentMarketPrice.multiply(getCurrentQuantity(bond)));
        }

        return total;
    }

    @Transactional(readOnly = true)
    public BigDecimal getInvestedValue(Bond bond) {
        return getAverageUnitPrice(bond).multiply(getCurrentQuantity(bond));
    }

    @Transactional(readOnly = true)
    public BigDecimal getTotalInterestReceived() {
        List<Bond> bonds = bondRepository.findByArchivedFalseOrderByNameAsc();
        BigDecimal total = BigDecimal.ZERO;

        for (Bond bond : bonds) {
            List<BondOperation> operations =
                    bondOperationRepository.findByBondOrderByOperationDateAsc(bond);

            for (BondOperation op : operations) {
                if (op.getOperationType() == OperationType.SELL && op.getNetProfit() != null) {
                    if (op.getNetProfit().compareTo(BigDecimal.ZERO) > 0) {
                        total = total.add(op.getNetProfit());
                    }
                }
            }
        }

        return total;
    }

    @Transactional(readOnly = true)
    public BigDecimal calculateOperationProfitLoss(BondOperation operation) {
        if (operation.getOperationType() == OperationType.BUY) {
            return BigDecimal.ZERO;
        }

        return operation.getNetProfit() != null ? operation.getNetProfit() : BigDecimal.ZERO;
    }

    @Transactional
    public void updateOperation(
            Integer operationId,
            Integer walletId,
            BigDecimal quantity,
            BigDecimal unitPrice,
            LocalDate operationDate,
            BigDecimal fees,
            BigDecimal taxes,
            BigDecimal netProfit,
            Category category,
            String description,
            TransactionStatus status) {
        BondOperation operation =
                bondOperationRepository
                        .findById(operationId)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                "BondOperation not found with id: " + operationId));

        if (quantity.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Quantity must be greater than zero");
        }

        if (unitPrice.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Unit price must be greater than zero");
        }

        if (operation.getOperationType() == OperationType.SELL) {
            BigDecimal currentQuantity = getCurrentQuantity(operation.getBond());
            BigDecimal availableQuantity = currentQuantity.add(operation.getQuantity());
            if (availableQuantity.compareTo(quantity) < 0) {
                throw new IllegalArgumentException(
                        "Insufficient bond quantity. Available: "
                                + availableQuantity
                                + ", Requested: "
                                + quantity);
            }
        }

        BigDecimal feesAmount = fees != null ? fees : BigDecimal.ZERO;
        BigDecimal taxesAmount = taxes != null ? taxes : BigDecimal.ZERO;
        BigDecimal netProfitAmount = netProfit != null ? netProfit : BigDecimal.ZERO;

        BigDecimal baseAmount = unitPrice.multiply(quantity);
        BigDecimal amount;

        if (operation.getOperationType() == OperationType.BUY) {
            amount = baseAmount.add(feesAmount).add(taxesAmount);
        } else {
            amount = baseAmount.add(netProfitAmount);
        }

        LocalDateTime dateTime = operationDate.atStartOfDay();

        WalletTransaction walletTransaction = operation.getWalletTransaction();
        walletTransaction.setWallet(walletService.getWalletById(walletId));
        walletTransaction.setCategory(category);
        walletTransaction.setDate(dateTime);
        walletTransaction.setAmount(amount);
        walletTransaction.setDescription(description);
        walletTransaction.setStatus(status);

        walletTransactionService.updateTransaction(walletTransaction);

        operation.setQuantity(quantity);
        operation.setUnitPrice(unitPrice);
        operation.setFees(fees);
        operation.setTaxes(taxes);
        operation.setNetProfit(netProfit);

        bondOperationRepository.save(operation);

        log.info("BondOperation with id {} updated successfully", operationId);
    }

    @Transactional
    public void deleteOperation(Integer operationId) {
        BondOperation operation =
                bondOperationRepository
                        .findById(operationId)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                "BondOperation not found with id: " + operationId));

        WalletTransaction walletTransaction = operation.getWalletTransaction();

        bondOperationRepository.delete(operation);

        if (walletTransaction != null) {
            walletTransactionService.deleteTransaction(walletTransaction.getId());
        }

        log.info("BondOperation with id {} deleted successfully", operationId);
    }

    @Transactional(readOnly = true)
    public BondOperation getOperationById(Integer operationId) {
        return bondOperationRepository
                .findById(operationId)
                .orElseThrow(
                        () ->
                                new EntityNotFoundException(
                                        "BondOperation not found with id: " + operationId));
    }
}

File: src/main/kotlin/org/moinex/service/BondService.kt
/*
 * Filename: BondService.kt (original filename: BondService.java)
 * Created on: January  2, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 11/03/2026
 */

package org.moinex.service

import org.moinex.common.extension.findByIdOrThrow
import org.moinex.common.extension.isZero
import org.moinex.common.extension.toRounded
import org.moinex.model.dto.BondOperationWalletTransactionDTO
import org.moinex.model.enums.OperationType
import org.moinex.model.enums.WalletTransactionType
import org.moinex.model.investment.Bond
import org.moinex.model.investment.BondInterestCalculation
import org.moinex.model.investment.BondOperation
import org.moinex.model.wallettransaction.WalletTransaction
import org.moinex.repository.investment.BondOperationRepository
import org.moinex.repository.investment.BondRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.time.LocalDateTime

@Service
class BondService(
    private val bondRepository: BondRepository,
    private val bondOperationRepository: BondOperationRepository,
    private val walletService: WalletService,
    private val bondInterestCalculationService: BondInterestCalculationService,
) {
    private val logger = LoggerFactory.getLogger(BondService::class.java)

    @Transactional
    fun createBond(bond: Bond) {
        bond.symbol?.let {
            check(!bondRepository.existsBySymbol(bond.symbol!!)) {
                "Bond with symbol ${bond.symbol} already exists"
            }
        }

        val newBond = bondRepository.save(bond)

        logger.info("$newBond created successfully")
    }

    @Transactional
    fun updateBond(updatedBond: Bond) {
        val bondFromDatabase = bondRepository.findByIdOrThrow(updatedBond.id!!)

        updatedBond.symbol?.let {
            check(!bondRepository.existsBySymbolAndIdNot(updatedBond.symbol!!, updatedBond.id!!)) {
                "Bond with symbol ${updatedBond.symbol} already exists"
            }
        }

        bondFromDatabase.apply {
            name = updatedBond.name
            symbol = updatedBond.symbol
            type = updatedBond.type
            issuer = updatedBond.issuer
            maturityDate = updatedBond.maturityDate
            interestType = updatedBond.interestType
            interestIndex = updatedBond.interestIndex
            interestRate = updatedBond.interestRate
        }

        logger.info("$bondFromDatabase updated successfully")
    }

    @Transactional
    fun deleteBond(id: Int) {
        val bondFromDatabase = bondRepository.findByIdOrThrow(id)

        val operations = bondOperationRepository.findByBondOrderByOperationDateAsc(bondFromDatabase)

        check(operations.isEmpty()) {
            "$bondFromDatabase has operations associated with it and cannot be deleted. " +
                "Remove the operations first or archive the bond"
        }

        bondRepository.delete(bondFromDatabase)

        logger.info("$bondFromDatabase deleted successfully")
    }

    @Transactional
    fun archiveBond(id: Int) {
        val bondFromDatabase = bondRepository.findByIdOrThrow(id)

        bondFromDatabase.archived = true

        logger.info("$bondFromDatabase archived successfully")
    }

    @Transactional
    fun unarchiveBond(id: Int) {
        val bondFromDatabase = bondRepository.findByIdOrThrow(id)

        bondFromDatabase.archived = false

        logger.info("$bondFromDatabase unarchived successfully")
    }

    @Transactional
    fun createBondOperation(
        bondOperation: BondOperation,
        bondOperationWalletTransactionDTO: BondOperationWalletTransactionDTO,
    ) {
        check(bondRepository.existsById(bondOperation.bond.id!!)) {
            "${bondOperation.bond} does not exist"
        }

        if (bondOperation.operationType == OperationType.SELL) {
            val currentQuantity = getCurrentQuantity(bondOperation.bond)

            check(currentQuantity >= bondOperation.quantity) {
                "Insufficient bond quantity. Available: $currentQuantity, Requested: ${bondOperation.quantity}"
            }
        }

        val baseAmount = bondOperation.unitPrice.multiply(bondOperation.quantity)
        val operationAmount =
            when (bondOperation.operationType) {
                OperationType.BUY -> baseAmount.add(bondOperation.fees).add(bondOperation.taxes)
                OperationType.SELL -> baseAmount.add(bondOperation.netProfit)
            }

        val transactionType =
            when (bondOperation.operationType) {
                OperationType.BUY -> WalletTransactionType.EXPENSE
                OperationType.SELL -> WalletTransactionType.INCOME
            }

        val transactionId =
            walletService.createWalletTransaction(
                WalletTransaction.from(bondOperationWalletTransactionDTO, transactionType, operationAmount),
            )

        val walletTransaction = walletService.getWalletTransactionById(transactionId)

        bondOperation.apply { this.walletTransaction = walletTransaction }

        bondOperationRepository.save(bondOperation)

        logger.info("$bondOperation created successfully")
    }

    @Transactional
    fun updateBondOperation(
        updatedBondOperation: BondOperation,
        bondOperationWalletTransactionDTO: BondOperationWalletTransactionDTO,
    ) {
        val bondOperationFromDatabase = bondOperationRepository.findByIdOrThrow(updatedBondOperation.id!!)

        if (bondOperationFromDatabase.operationType == OperationType.SELL) {
            val currentQuantity = getCurrentQuantity(bondOperationFromDatabase.bond)
            val availableQuantity = currentQuantity.add(bondOperationFromDatabase.quantity)

            check(availableQuantity >= updatedBondOperation.quantity) {
                "Insufficient bond quantity. Available: $availableQuantity, Requested: ${updatedBondOperation.quantity}"
            }
        }

        val baseAmount = updatedBondOperation.unitPrice.multiply(updatedBondOperation.quantity)
        val operationAmount =
            when (bondOperationFromDatabase.operationType) {
                OperationType.BUY -> baseAmount.add(updatedBondOperation.fees).add(updatedBondOperation.taxes)
                OperationType.SELL -> baseAmount.add(updatedBondOperation.netProfit)
            }

        val walletTransaction = bondOperationFromDatabase.walletTransaction!!

        walletTransaction.apply {
            wallet = bondOperationWalletTransactionDTO.wallet
            category = bondOperationWalletTransactionDTO.category
            date = bondOperationWalletTransactionDTO.date
            amount = operationAmount
            description = bondOperationWalletTransactionDTO.description
            status = bondOperationWalletTransactionDTO.status
            includeInAnalysis = bondOperationWalletTransactionDTO.includeInAnalysis
        }

        walletService.updateWalletTransaction(walletTransaction)

        bondOperationFromDatabase
            .apply {
                quantity = updatedBondOperation.quantity
                unitPrice = updatedBondOperation.unitPrice
                fees = updatedBondOperation.fees
                taxes = updatedBondOperation.taxes
            }.also {
                if (it.operationType == OperationType.SELL) {
                    it.netProfit = updatedBondOperation.netProfit
                }
            }

        logger.info("$bondOperationFromDatabase updated successfully")
    }

    @Transactional
    fun deleteBondOperation(operationId: Int) {
        val bondOperationFromDatabase = bondOperationRepository.findByIdOrThrow(operationId)

        val walletTransaction = bondOperationFromDatabase.walletTransaction!!

        bondOperationRepository.delete(bondOperationFromDatabase)

        walletService.deleteWalletTransaction(walletTransaction.id!!)

        logger.info("$bondOperationFromDatabase deleted successfully")
    }

    @Transactional(readOnly = true)
    fun getCurrentQuantity(bond: Bond): BigDecimal {
        val operations = bondOperationRepository.findByBondOrderByOperationDateAsc(bond)

        return operations.fold(BigDecimal.ZERO) { quantity, op ->
            when (op.operationType) {
                OperationType.BUY -> quantity.add(op.quantity)
                OperationType.SELL -> quantity.subtract(op.quantity)
            }
        }
    }

    @Transactional(readOnly = true)
    fun getAverageUnitPrice(bond: Bond): BigDecimal {
        val purchases =
            bondOperationRepository.findByBondAndOperationTypeOrderByOperationDateAsc(
                bond,
                OperationType.BUY,
            )

        if (purchases.isEmpty()) {
            return BigDecimal.ZERO
        }

        val totalValue =
            purchases.fold(BigDecimal.ZERO) { acc, purchase ->
                acc.add(purchase.unitPrice.multiply(purchase.quantity))
            }

        val totalQuantity =
            purchases.fold(BigDecimal.ZERO) { acc, purchase ->
                acc.add(purchase.quantity)
            }

        if (totalQuantity.isZero()) {
            return BigDecimal.ZERO
        }

        return totalValue.divide(totalQuantity).toRounded()
    }

    @Transactional(readOnly = true)
    fun getAllNonArchivedBonds(): List<Bond> = bondRepository.findByArchivedFalseOrderByNameAsc()

    @Transactional(readOnly = true)
    fun getAllArchivedBonds(): List<Bond> = bondRepository.findByArchivedTrueOrderByNameAsc()

    @Transactional(readOnly = true)
    fun getBondById(id: Int): Bond = bondRepository.findByIdOrThrow(id)

    @Transactional(readOnly = true)
    fun getAllOperations(): List<BondOperation> = bondOperationRepository.findAllByOrderByOperationDateDesc()

    @Transactional(readOnly = true)
    fun getOperationsByBond(bond: Bond): List<BondOperation> = bondOperationRepository.findByBondOrderByOperationDateAsc(bond)

    @Transactional(readOnly = true)
    fun getTotalProfit(bond: Bond): BigDecimal {
        val operations = bondOperationRepository.findByBondAndOperationTypeOrderByOperationDateAsc(bond, OperationType.SELL)

        return operations.sumOf { it.netProfit }
    }

    @Transactional(readOnly = true)
    fun getTotalInvestedValue(): BigDecimal {
        val bonds = bondRepository.findByArchivedFalseOrderByNameAsc()

        return bonds
            .filter { getCurrentQuantity(it) > BigDecimal.ZERO }
            .fold(BigDecimal.ZERO) { total, bond ->
                val currentQuantity = getCurrentQuantity(bond)
                val averagePrice = getAverageUnitPrice(bond)
                total.add(averagePrice.multiply(currentQuantity))
            }
    }

    @Transactional(readOnly = true)
    fun getTotalCurrentValue(currentMarketPrice: BigDecimal): BigDecimal {
        val bonds = bondRepository.findByArchivedFalseOrderByNameAsc()

        return bonds.fold(BigDecimal.ZERO) { total, bond ->
            total.add(currentMarketPrice.multiply(getCurrentQuantity(bond)))
        }
    }

    @Transactional(readOnly = true)
    fun getTotalInvestedValue(bond: Bond): BigDecimal = getAverageUnitPrice(bond).multiply(getCurrentQuantity(bond))

    @Transactional(readOnly = true)
    fun getAllBondsTotalAccumulatedInterest(): BigDecimal {
        val bonds = bondRepository.findByArchivedFalseOrderByNameAsc()

        return bonds.sumOf { getTotalAccumulatedInterestByBondId(it.id!!) }
    }

    @Transactional(readOnly = true)
    fun getOperationCountByBond(bondId: Int): Int {
        val bondFromDatabase = bondRepository.findByIdOrThrow(bondId)

        return bondOperationRepository.findByBondOrderByOperationDateAsc(bondFromDatabase).size
    }

    @Transactional(readOnly = true)
    fun getOperationsByDateBefore(date: LocalDateTime): List<BondOperation> = bondOperationRepository.findAllByDateBefore(date)

    @Transactional(readOnly = true)
    fun getMonthlyInterestHistory(bondId: Int): List<BondInterestCalculation> {
        val bondFromDatabase = bondRepository.findByIdOrThrow(bondId)
        return bondInterestCalculationService.getMonthlyInterestHistory(bondFromDatabase)
    }

    @Transactional(readOnly = true)
    fun getCurrentMonthInterest(bondId: Int): BigDecimal =
        bondInterestCalculationService
            .getCurrentMonthInterest(bondId) ?: BigDecimal.ZERO

    @Transactional(readOnly = true)
    fun getTotalAccumulatedInterestByBondId(bondId: Int): BigDecimal =
        bondInterestCalculationService.getLatestCalculation(bondId)?.accumulatedInterest ?: BigDecimal.ZERO
}
